File: exercism/elixir/protein-translation/lib/protein_translation.ex
defmodule ProteinTranslation do
  @codon2protein %{
    "UGU" => "Cysteine",
    "UGC" => "Cysteine",
    "UUA" => "Leucine",
    "UUG" => "Leucine",
    "AUG" => "Methionine",
    "UUU" => "Phenylalanine",
    "UUC" => "Phenylalanine",
    "UCU" => "Serine",
    "UCC" => "Serine",
    "UCA" => "Serine",
    "UCG" => "Serine",
    "UGG" => "Tryptophan",
    "UAU" => "Tyrosine",
    "UAC" => "Tyrosine",
    "UAA" => "STOP",
    "UAG" => "STOP",
    "UGA" => "STOP"
  }

  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {:ok, list(String.t())} | {:error, String.t()}
  def of_rna(rna) do
    rna
    |> String.split("", trim: true)
    |> Enum.chunk_every(3)
    |> Enum.map(&Enum.join/1)
    |> do_of_rna([])
  end

  defp do_of_rna([], acc), do: {:ok, Enum.reverse(acc)}

  defp do_of_rna([h | t], acc) do
    case of_codon(h) do
      {:error, _} -> {:error, "invalid RNA"}
      {:ok, "STOP"} -> do_of_rna([], acc)
      {:ok, protein} -> do_of_rna(t, [protein | acc])
    end
  end

  @doc """
  Given a codon, return the corresponding protein


  """
  @spec of_codon(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def of_codon(codon) do
    case Map.get(@codon2protein, codon) do
      nil -> {:error, "invalid codon"}
      protein -> {:ok, protein}
    end
  end
end

File: exercism/haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where

proteins :: String -> Maybe [String]
proteins = proteins' []

proteins' :: [String] -> String -> Maybe [String]
proteins' acc rna =
  case codon rna of
    Just (x, xs) ->
      case translate x of
        Just protein -> proteins' (protein:acc) xs
        Nothing -> Just $ reverse acc
    Nothing -> Just $ reverse acc

codon :: String -> Maybe (String, String)
codon (x1:x2:x3:xs) = Just ([x1, x2, x3], xs)
codon _ = Nothing

translate :: String -> Maybe String
translate codon
  | codon `elem` ["AUG"] = Just "Methionine"
  | codon `elem` ["UUU", "UUC"] = Just "Phenylalanine"
  | codon `elem` ["UUA", "UUG"] = Just "Leucine"
  | codon `elem` ["UCU", "UCC", "UCA", "UCG"] = Just "Serine"
  | codon `elem` ["UAU", "UAC"] = Just "Tyrosine"
  | codon `elem` ["UGU", "UGC"] = Just "Cysteine"
  | codon `elem` ["UGG"] = Just "Tryptophan"
  | codon `elem` ["UAA", "UAG", "UGA"] = Nothing